sphinx/util/fileutil.py

```diff
@@ -4,17 +4,33 @@
 
 import os
 import posixpath
-from typing import TYPE_CHECKING, Callable
+import warnings
+from typing import TYPE_CHECKING, Callable, Optional
 
 from docutils.utils import relative_path
 
+from sphinx.deprecation import RemovedInSphinx80Warning
 from sphinx.util.osutil import copyfile, ensuredir
 from sphinx.util.typing import PathMatcher
 
 if TYPE_CHECKING:
     from sphinx.util.template import BaseRenderer
 
 
+def _template_basename(filename: str) -> Optional[str]:
```

Comment:
`Optional` is used here to support Python 3.8.  Ruff wants to suggest `str | None`, but doesn't appear to limit that suggestion based on [configured Python version](https://github.com/sphinx-doc/sphinx/blob/ecfd08d3254587972299c23cd596660f5796099c/pyproject.toml#L140) ([union type syntax is available from Python 3.10](https://docs.python.org/3.10/whatsnew/3.10.html#pep-604-new-type-union-operator)) - I'll report that to `ruff`.

Reply:
Nope, my mistake; that suggestion is [valid when `from __future__ import annotations` is in use](https://github.com/charliermarsh/ruff/issues/2382#issuecomment-1409939445).  Today I learned.